src/compiler/binding.ts

```diff
@@ -351,147 +334,273 @@ export function concatNamesAndValues(
+				);
+			}
+		} else if (ts.TypeGuards.isPropertyAssignment(property)) {
+			const name = property.getNameNode();
+			const initializer = property.getInitializerOrThrow();
+			const rhs = objectAccessor(state, parentId, name, getAccessor, name, name);
+			if (
+				ts.TypeGuards.isIdentifier(initializer) ||
+				ts.TypeGuards.isElementAccessExpression(initializer) ||
+				ts.TypeGuards.isPropertyAccessExpression(initializer)
+			) {
+				const nameStr = compileExpression(state, initializer);
+				state.pushPrecedingStatements(bindingLiteral, state.indent + `${nameStr} = ${rhs};\n`);
+			} else if (ts.TypeGuards.isBinaryExpression(initializer)) {
+				const nameStr = compileExpression(state, skipNodesDownwards(initializer.getLeft()));
+				state.pushPrecedingStatements(bindingLiteral, state.indent + `${nameStr} = ${rhs};\n`);
+				const init = skipNodesDownwards(initializer.getRight());
+				state.pushPrecedingStatements(
+					init,
+					state.indent + compileParamDefault(state, initializer, nameStr) + "\n",
+				);
+			} else if (
+				ts.TypeGuards.isObjectLiteralExpression(initializer) ||
+				ts.TypeGuards.isArrayLiteralExpression(initializer)
+			) {
+				const id = state.getNewId();
+				state.pushPrecedingStatements(bindingLiteral, state.indent + `local ${id} = ${rhs};\n`);
+				compileBindingLiteralInner(state, initializer, id, accessNode);
 			}
-		} else if (ts.TypeGuards.isOmittedExpression(item)) {
-			getAccessor(state, parentId, childIndex, preStatements, idStack, true);
 		} else {
 			throw new CompilerError(
-				`Unexpected ${item.getKindName()} in getBindingData.`,
-				item,
+				`Unexpected ${property.getKindName()} in compileArrayBindingLiteral.`,
+				property,
 				CompilerErrorType.UnexpectedBindingPattern,
 				true,
 			);
 		}
+	}
+}
 
-		childIndex++;
+function compileBindingLiteralInner(
+	state: CompilerState,
+	bindingLiteral: BindingLiteral,
+	parentId: string,
+	accessNode: ts.Node,
+) {
+	if (ts.TypeGuards.isArrayLiteralExpression(bindingLiteral)) {
+		compileArrayBindingLiteral(state, bindingLiteral, parentId, accessNode);
+	} else if (ts.TypeGuards.isObjectLiteralExpression(bindingLiteral)) {
+		compileObjectBindingLiteral(state, bindingLiteral, parentId, accessNode);
 	}
 }
+
+export function compileBindingLiteralAndJoin(
+	state: CompilerState,
+	bindingLiteral: BindingLiteral,
+	parentId: string,
+	accessNode: ts.Node = bindingLiteral,
+) {
+	state.enterPrecedingStatementContext();
+	compileBindingLiteralInner(state, bindingLiteral, parentId, accessNode);
+	return state.exitPrecedingStatementContextAndJoin();
+}
+
+export function compileBindingLiteral(
+	state: CompilerState,
+	bindingLiteral: BindingLiteral,
+	parentId: string,
+	accessNode: ts.Node = bindingLiteral,
+) {
+	state.enterPrecedingStatementContext();
+	compileBindingLiteralInner(state, bindingLiteral, parentId, accessNode);
+	return state.exitPrecedingStatementContext().map(v => v.trim());
```

Comment:
Why is this trimmed? This array of strings could be of anything, and there are no guarentees about how joined statements are. This could break whitespace either now or sometime in the future.

Reply:
Wasn't super sure on this part. It's because I have `state.indent` and `\n` before/after each statement I'm pushing, but the call sites are not expecting those.

I guess I could just rework the call sites?

Reply:
I did find that I _needed_ the whitespace pushed for the join version.